File: ios/ios/PreAuth/PreAuthViewModel.swift
//
//  PreAuthViewModel.swift
//  ios
//
//  Pre-auth flow - matches APK PreAuthViewModel
//

import Foundation

enum PreAuthStep {
    case splash
    case updateRequired
    case deviceVerification
    case onboarding
    case termsConditions
    case privacyPolicy
    case dataProcessingConsent
    case permissionExplainer
    case auth
}

@MainActor
@Observable
class PreAuthViewModel {
    var step: PreAuthStep = .splash
    var isLoading = true
    var forceUpdate = false
    var appStoreUrl = "https://apps.apple.com/app/monytix"
    var securityWarning = false

    private let api = BackendApi()

    init() {
        Task { await runSplashChecks() }
    }

    func runSplashChecks() async {
        isLoading = true
        try? await Task.sleep(nanoseconds: 500_000_000) // 500ms min splash

        do {
            let config = try await api.getConfig()
            let versionCode = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "1"
            let currentVersion = Int(versionCode) ?? 1
            let minRequired = config.minVersionCode ?? 1
            forceUpdate = currentVersion < minRequired
            appStoreUrl = config.appStoreUrl ?? "https://apps.apple.com/app/monytix"
        } catch {
            // Timeout or failure: proceed with defaults
            forceUpdate = false
            appStoreUrl = "https://apps.apple.com/app/monytix"
        }

        // iOS: no test-keys / root detection like Android; skip device verification
        let securityOk = true
        isLoading = false
        securityWarning = !securityOk
        await advanceToNextStep(updateRequired: forceUpdate, securityOk: securityOk)
    }

    private func advanceToNextStep(updateRequired: Bool, securityOk: Bool) async {
        if updateRequired {
            step = .updateRequired
            return
        }
        if !securityOk {
            step = .deviceVerification
            return
        }
        if !PreAuthPreferences.onboardingComplete {
            step = .onboarding
            return
        }
        if !PreAuthPreferences.termsAccepted {
            step = .termsConditions
            return
        }
        if !PreAuthPreferences.privacyViewed {
            step = .privacyPolicy
            return
        }
        if !PreAuthPreferences.dataConsent {
            step = .dataProcessingConsent
            return
        }
        if !PreAuthPreferences.permissionsExplained {
            step = .permissionExplainer
            return
        }
        step = .auth
    }

    func completeOnboarding() {
        AnalyticsHelper.logEvent("onboarding_complete")
        PreAuthPreferences.onboardingComplete = true
        step = .termsConditions
    }

    func acceptTerms() {
        AnalyticsHelper.logEvent("terms_accepted")
        PreAuthPreferences.termsAccepted = true
        step = .privacyPolicy
    }

    func completePrivacyPolicy() {
        AnalyticsHelper.logEvent("privacy_continue")
        PreAuthPreferences.privacyViewed = true
        step = .dataProcessingConsent
    }

    func acceptDataConsent() {
        AnalyticsHelper.logEvent("data_consent_accepted")
        PreAuthPreferences.dataConsent = true
        step = .permissionExplainer
    }

    func completePermissionExplainer() {
        AnalyticsHelper.logEvent("permission_continue")
        PreAuthPreferences.permissionsExplained = true
        step = .auth
    }

    func completeDeviceVerification() {
        AnalyticsHelper.logEvent("device_verification_continue")
        Task {
            await advanceToNextStep(updateRequired: false, securityOk: true)
        }
    }

    func goToAuth() {
        AnalyticsHelper.logEvent("login_clicked")
        PreAuthPreferences.setAllComplete()
        step = .auth
    }
}

File: apk copy/app/src/main/java/com/example/monytix/preauth/PreAuthViewModel.kt
package com.example.monytix.preauth

import android.app.Application
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.monytix.BuildConfig
import com.example.monytix.analytics.AnalyticsHelper
import com.example.monytix.data.BackendApi
import kotlinx.coroutines.delay
import kotlinx.coroutines.withTimeoutOrNull
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

sealed class PreAuthStep {
    data object Splash : PreAuthStep()
    data object UpdateRequired : PreAuthStep()
    data object DeviceVerification : PreAuthStep()
    data object Onboarding : PreAuthStep()
    data object TermsConditions : PreAuthStep()
    data object PrivacyPolicy : PreAuthStep()
    data object DataProcessingConsent : PreAuthStep()
    data object PermissionExplainer : PreAuthStep()
    data object Auth : PreAuthStep()
}

data class PreAuthUiState(
    val step: PreAuthStep = PreAuthStep.Splash,
    val isLoading: Boolean = true,
    val forceUpdate: Boolean = false,
    val appStoreUrl: String = "",
    val securityWarning: Boolean = false,
)

class PreAuthViewModel(
    private val application: Application
) : ViewModel() {

    private val _uiState = MutableStateFlow(PreAuthUiState())
    val uiState: StateFlow<PreAuthUiState> = _uiState.asStateFlow()

    init {
        runSplashChecks()
    }

    private fun runSplashChecks() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            delay(500) // Minimum splash display time
            val configResult = withTimeoutOrNull(5_000L) { BackendApi.getConfig() } ?: run {
                // Timeout or failure: proceed with defaults (no force update)
                Result.success(
                    com.example.monytix.data.ConfigResponse(
                        min_version_code = 1,
                        app_store_url = "https://play.google.com/store/apps/details?id=com.example.monytix",
                        feature_flags = emptyMap(),
                        maintenance_mode = false
                    )
                )
            }
            val updateRequired = configResult.getOrNull()?.let { config ->
                BuildConfig.VERSION_CODE < config.min_version_code
            } ?: false
            val appStoreUrl = configResult.getOrNull()?.app_store_url ?: ""
            val securityOk = !isDeviceCompromised()
            _uiState.update {
                it.copy(
                    isLoading = false,
                    forceUpdate = updateRequired,
                    appStoreUrl = appStoreUrl,
                    securityWarning = !securityOk
                )
            }
            advanceToNextStep(updateRequired, securityOk)
        }
    }

    private suspend fun advanceToNextStep(updateRequired: Boolean, securityOk: Boolean) {
        if (updateRequired) {
            _uiState.update { it.copy(step = PreAuthStep.UpdateRequired) }
            return
        }
        if (!securityOk) {
            _uiState.update { it.copy(step = PreAuthStep.DeviceVerification) }
            return
        }
        val onboardingDone = PreAuthPreferences.onboardingComplete(application).first()
        if (!onboardingDone) {
            _uiState.update { it.copy(step = PreAuthStep.Onboarding) }
            return
        }
        val termsAccepted = PreAuthPreferences.termsAccepted(application).first()
        if (!termsAccepted) {
            _uiState.update { it.copy(step = PreAuthStep.TermsConditions) }
            return
        }
        val privacyViewed = PreAuthPreferences.privacyViewed(application).first()
        if (!privacyViewed) {
            _uiState.update { it.copy(step = PreAuthStep.PrivacyPolicy) }
            return
        }
        val dataConsent = PreAuthPreferences.dataConsent(application).first()
        if (!dataConsent) {
            _uiState.update { it.copy(step = PreAuthStep.DataProcessingConsent) }
            return
        }
        val permissionsExplained = PreAuthPreferences.permissionsExplained(application).first()
        if (!permissionsExplained) {
            _uiState.update { it.copy(step = PreAuthStep.PermissionExplainer) }
            return
        }
        _uiState.update { it.copy(step = PreAuthStep.Auth) }
    }

    private fun isDeviceCompromised(): Boolean {
        return try {
            val buildTags = android.os.Build.TAGS
            buildTags != null && buildTags.contains("test-keys")
        } catch (_: Exception) {
            false
        }
    }

    fun completeOnboarding() {
        viewModelScope.launch {
            AnalyticsHelper.logEvent("onboarding_complete")
            PreAuthPreferences.setOnboardingComplete(application, true)
            _uiState.update { it.copy(step = PreAuthStep.TermsConditions) }
        }
    }

    fun acceptTerms() {
        viewModelScope.launch {
            AnalyticsHelper.logEvent("terms_accepted")
            PreAuthPreferences.setTermsAccepted(application, true)
            _uiState.update { it.copy(step = PreAuthStep.PrivacyPolicy) }
        }
    }

    fun completePrivacyPolicy() {
        viewModelScope.launch {
            AnalyticsHelper.logEvent("privacy_continue")
            PreAuthPreferences.setPrivacyViewed(application, true)
            _uiState.update { it.copy(step = PreAuthStep.DataProcessingConsent) }
        }
    }

    fun acceptDataConsent() {
        viewModelScope.launch {
            AnalyticsHelper.logEvent("data_consent_accepted")
            PreAuthPreferences.setDataConsent(application, true)
            _uiState.update { it.copy(step = PreAuthStep.PermissionExplainer) }
        }
    }

    fun completePermissionExplainer() {
        viewModelScope.launch {
            AnalyticsHelper.logEvent("permission_continue")
            PreAuthPreferences.setPermissionsExplained(application, true)
            _uiState.update { it.copy(step = PreAuthStep.Auth) }
        }
    }

    fun skipOnboarding() = completeOnboarding()

    fun completeDeviceVerification() {
        viewModelScope.launch {
            AnalyticsHelper.logEvent("device_verification_continue")
            advanceToNextStep(updateRequired = false, securityOk = true)
        }
    }

    fun goToAuth() {
        viewModelScope.launch {
            AnalyticsHelper.logEvent("login_clicked")
            PreAuthPreferences.setOnboardingComplete(application, true)
            PreAuthPreferences.setTermsAccepted(application, true)
            PreAuthPreferences.setPrivacyViewed(application, true)
            PreAuthPreferences.setDataConsent(application, true)
            PreAuthPreferences.setPermissionsExplained(application, true)
            _uiState.update { it.copy(step = PreAuthStep.Auth) }
        }
    }

    /** Go to the previous step in the pre-auth flow. No-op for Splash, UpdateRequired, DeviceVerification, Onboarding. */
    fun goBack() {
        val previous = when (_uiState.value.step) {
            PreAuthStep.Auth -> PreAuthStep.PermissionExplainer
            PreAuthStep.PermissionExplainer -> PreAuthStep.DataProcessingConsent
            PreAuthStep.DataProcessingConsent -> PreAuthStep.PrivacyPolicy
            PreAuthStep.PrivacyPolicy -> PreAuthStep.TermsConditions
            PreAuthStep.TermsConditions -> PreAuthStep.Onboarding
            else -> return
        }
        _uiState.update { it.copy(step = previous) }
    }
}
